timelock-agent/src/main/java/com/palantir/timelock/paxos/BatchTimelockPaxosLearnerAdapter.java

```diff
@@ -0,0 +1,108 @@
+import java.util.Set;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.timelock.paxos.BatchPaxosLearnerRpcClient;
+import com.palantir.atlasdb.timelock.paxos.Client;
+import com.palantir.atlasdb.timelock.paxos.PaxosUseCase;
+import com.palantir.atlasdb.timelock.paxos.WithSeq;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosValue;
+
+public class BatchTimelockPaxosLearnerAdapter implements PaxosLearner {
+    private final PaxosUseCase paxosUseCase;
+    private final Client client;
+    private final BatchPaxosLearnerRpcClient rpcClient;
+
+    private volatile long lastKnownSequence = Long.MIN_VALUE;
+
+    public BatchTimelockPaxosLearnerAdapter(
+            PaxosUseCase paxosUseCase,
+            Client client,
+            BatchPaxosLearnerRpcClient rpcClient) {
+        this.paxosUseCase = paxosUseCase;
+        this.client = client;
+        this.rpcClient = rpcClient;
+    }
+
+    public static PaxosLearner singleLeader(BatchPaxosLearnerRpcClient rpcClient) {
+        return new BatchTimelockPaxosLearnerAdapter(
+                PaxosUseCase.LEADER_FOR_ALL_CLIENTS,
+                PaxosUseCase.PSEUDO_LEADERSHIP_CLIENT,
+                rpcClient);
+    }
+
+    @Override
+    public void learn(long seq, PaxosValue val) {
+        rpcClient.learn(paxosUseCase, ImmutableSetMultimap.of(client, val));
+        updateLastKnownSequence(seq);
+    }
+
+    @Override
+    public Optional<PaxosValue> getLearnedValue(long seq) {
+        Set<PaxosValue> result = rpcClient.getLearnedValues(paxosUseCase, ImmutableSet.of(WithSeq.of(client, seq)))
+                .get(client);
+        checkResult(result);
+        if (result.isEmpty()) {
+            return Optional.empty();
+        }
+        updateLastKnownSequence(seq);
+        return Optional.of(Iterables.getOnlyElement(result));
+    }
+
+    @Override
+    public Optional<PaxosValue> getGreatestLearnedValue() {
+        Set<PaxosValue> result = rpcClient
+                .getLearnedValuesSince(paxosUseCase, ImmutableMap.of(client, lastKnownSequence))
+                .get(client);
+        Optional<PaxosValue> greatestIfExists = result.stream()
+                .reduce((fst, snd) -> fst.getRound() > snd.getRound() ? fst : snd);
+        greatestIfExists.ifPresent(paxosValue -> updateLastKnownSequence(paxosValue.getRound()));
+        return greatestIfExists;
+    }
+
+    @Override
+    public Collection<PaxosValue> getLearnedValuesSince(long seq) {
+        Set<PaxosValue> result = rpcClient.getLearnedValuesSince(paxosUseCase, ImmutableMap.of(client, seq))
+                .get(client);
+        result.stream().map(PaxosValue::getRound).mapToLong(x -> x).max().ifPresent(this::updateLastKnownSequence);
```

Comment:
Nice! Many ways to go about this, and looks like you picked the one that touches the volatile/atomic thing as little as possible.